### PR TITLE
Bootstrapped regression bounds

### PR DESCRIPTION
**Pull request by paxcema**
## Why

To offer an alternative method for regression bounds.

## How

Introducing a `PLinearWrapper` class for the (already existing) implementation of `PLinear` layers by @torrmal. This enables its usage as a post-training analysis block where, essentially, the network will learn to predict distribution parameters to model the residual error for any given (input, prediction) tuple.

At inference time, these parameters are used to generate multiple samples that (when sorted) can enable quantile-based prediction intervals. This is not as ideal as simply optimizing the pinball loss directly in the underlying model, but model agnosticity is a pretty important thing to uphold, and this is my best attempt at that. Of course, benchmarks will have to confirm whether this approach is any good.

### CODE REVIEW

File: lightwood/analysis/nn_conf/plinear.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        conf_level = ns.pred_args.fixed_confidence
+        if conf_level is None:
+            conf_level = 0.95
+
+        X = ns.encoded_data
+        yh = torch.tensor(ns.predictions['prediction'].values).unsqueeze(1)
+        X = torch.hstack([X, yh]).float()
+
+        samples = []
+        for _ in range(self.n_samples):
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/analysis/nn_conf/plinear.py:91]: Just tried this out. There is a steep memory-time trade off here, because samples are not in the batch dimension: each data point in the batch should be sampled `n_samples` times. 

We can either increase the size of the batch `n_samples` times and manually pack/unpack the result (which could get out of hand for big datasets where the validation dataset is `10^5` or higher), or keep the iterator.

Only because we are not currently enforcing a maximum validation dataset size, I'm tempted to keep the current approach to avoid OOM issues. Thoughts?

File: lightwood/analysis/nn_conf/plinear.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            samples.append(self.model.forward(X))
+
+        samples = torch.stack(samples).squeeze()
+        sorted_samples = torch.sort(samples, dim=0).values
+        sorted_samples = sorted_samples - sorted_samples.median(dim=0).values  # rescale so that median is null
+        min_bound = sorted_samples[round(samples.shape[0] * (1.0 - conf_level)), :]
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/analysis/nn_conf/plinear.py]: Ok, I've added a secondary mode so that we can benchmark both the quantile methodology above, or a standard computation of the confidence interval assuming normal distribution of the residuals.

**Comment by paxcema** [comment]
Converting to draft as there is work pending to make this a fully sample-aware method.

**Comment by paxcema** [comment]
Closing for now.
